File: src/test_programs/C/test_SI_geomgrad.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "openmmpol.h"

void print_qmmm_grad(char *name, int32_t mm_atoms, int32_t qm_atoms, 
                     double *grad_mm, double *grad_qm){
    char msg[OMMP_STR_CHAR_MAX];
    
    sprintf(msg, "Grad %s", name);
    ommp_message(msg, OMMP_VERBOSE_NONE, "TEST-GRD");
    
    for(int i = 0; i < mm_atoms; i++){
        sprintf(msg, "MM:%-8d %+20.8g %+20.8g %+20.8g", i,
                grad_mm[i*3+0]*OMMP_AU2KCALMOL*OMMP_ANG2AU,
                grad_mm[i*3+1]*OMMP_AU2KCALMOL*OMMP_ANG2AU,
                grad_mm[i*3+2]*OMMP_AU2KCALMOL*OMMP_ANG2AU);
        ommp_message(msg, OMMP_VERBOSE_NONE, "TEST-GRD");
    }
    if(grad_qm != NULL && qm_atoms > 0){
        for(int i = 0; i < qm_atoms; i++){
            sprintf(msg, "QM:%-8d %+20.8g %+20.8g %+20.8g", i,
                    grad_qm[i*3+0]*OMMP_AU2KCALMOL*OMMP_ANG2AU,
                    grad_qm[i*3+1]*OMMP_AU2KCALMOL*OMMP_ANG2AU,
                    grad_qm[i*3+2]*OMMP_AU2KCALMOL*OMMP_ANG2AU);
            ommp_message(msg, OMMP_VERBOSE_NONE, "TEST-GRD");
        }
    }
    ommp_message("", OMMP_VERBOSE_NONE, "TEST-GRD");
}

void ana_grd_print(OMMP_SYSTEM_PRT sys,
                  void (*grad_f)(OMMP_SYSTEM_PRT, double *),
                  const char *name){
    
    int mm_atoms = ommp_get_mm_atoms(sys);
    
    double *_grad_ana = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    double **grad_ana = (double **) malloc(sizeof(double *) * mm_atoms);
    char msg[OMMP_STR_CHAR_MAX];

    for(int i = 0; i < mm_atoms; i++){
        grad_ana[i] = &(_grad_ana[i*3]);
        for(int j=0; j < 3; j++)
            grad_ana[i][j] = 0.0;
    }

    grad_f(sys, _grad_ana);
   
    print_qmmm_grad(name, mm_atoms, 0, _grad_ana, NULL);
}

void ana_grd_print_qmmm(OMMP_SYSTEM_PRT sys,
                        OMMP_QM_HELPER_PRT qmh,
                        OMMP_SYSTEM_PRT qm_fake,
                        void (*grad_f)(OMMP_SYSTEM_PRT, 
                                       OMMP_QM_HELPER_PRT, 
                                       OMMP_SYSTEM_PRT,
                                       double *, double *),
                        const char *name){
    char msg[OMMP_STR_CHAR_MAX];
    
    int mm_atoms = ommp_get_mm_atoms(sys);
    double *_grad_ana_mm = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    double **grad_ana_mm = (double **) malloc(sizeof(double *) * mm_atoms);
    for(int i = 0; i < mm_atoms; i++){
        grad_ana_mm[i] = &(_grad_ana_mm[i*3]);
        for(int j=0; j < 3; j++)
            grad_ana_mm[i][j] = 0.0;
    }
    
    int qm_atoms = ommp_qm_helper_get_qm_atoms(qmh);
    double *_grad_ana_qm = (double *) malloc(sizeof(double) * 3 * qm_atoms);
    double **grad_ana_qm = (double **) malloc(sizeof(double *) * qm_atoms);
    for(int i = 0; i < qm_atoms; i++){
        grad_ana_qm[i] = &(_grad_ana_qm[i*3]);
        for(int j=0; j < 3; j++)
            grad_ana_qm[i][j] = 0.0;
    }

    grad_f(sys, qmh, qm_fake, _grad_ana_mm, _grad_ana_qm);
   
    print_qmmm_grad(name, mm_atoms, qm_atoms, _grad_ana_mm, _grad_ana_qm);
}

void ommptest_qm_helper_vdw_geomgrad(OMMP_SYSTEM_PRT sys, OMMP_QM_HELPER_PRT qmh,
                                     OMMP_SYSTEM_PRT fakeqm, double *mmg, double *qmg){
    // Just for signature compliance
    ommp_qm_helper_vdw_geomgrad(qmh, sys, qmg, mmg);
}

void ommptest_fakeqm_internal_geomgrad(OMMP_SYSTEM_PRT sys, OMMP_QM_HELPER_PRT qmh,
                                     OMMP_SYSTEM_PRT fakeqm, double *mmg, double *qmg){
    // Compute derivatives of internal "QM" geometry under the assumption that MM 
    // electric field does not interact with "QM" static multipoles
    ommp_full_geomgrad(fakeqm, qmg);
}

void ommptest_fakeqm_linkatom_geomgrad(OMMP_SYSTEM_PRT sys, OMMP_QM_HELPER_PRT qmh,
                                       OMMP_SYSTEM_PRT fakeqm, double *mmg, double *qmg){
    int qm_atoms = ommp_qm_helper_get_qm_atoms(qmh);
    int mm_atoms = ommp_get_mm_atoms(sys);
    double *oldgrd = (double *) malloc(sizeof(double) * 3 * qm_atoms);
    ommp_qm_helper_vdw_geomgrad(qmh, sys, qmg, mmg);
    ommp_full_geomgrad(fakeqm, oldgrd);
    for(int i=0; i < 3 * qm_atoms; i++) oldgrd[i] += qmg[i]; // Remember, array are reset
    ommp_qm_helper_link_atom_geomgrad(qmh, sys, qmg, mmg, oldgrd);
    free(oldgrd);
}

void ommptest_totalqmmm_geomgrad(OMMP_SYSTEM_PRT sys, OMMP_QM_HELPER_PRT qmh,
                                 OMMP_SYSTEM_PRT fakeqm, double *mmg, double *qmg){
    int qm_atoms = ommp_qm_helper_get_qm_atoms(qmh);
    int mm_atoms = ommp_get_mm_atoms(sys);

    double *oldgrd_mm = (double *) malloc(sizeof(double) * 3 * mm_atoms);
    double *oldgrd_qm = (double *) malloc(sizeof(double) * 3 * qm_atoms);
    
    if(qmh != NULL) ommp_qm_helper_vdw_geomgrad(qmh, sys, oldgrd_qm, oldgrd_mm);
    for(int i=0; i < 3 * qm_atoms; i++) qmg[i] += oldgrd_qm[i];
    for(int i=0; i < 3 * mm_atoms; i++) mmg[i] += oldgrd_mm[i];
    if(fakeqm != NULL) ommp_full_geomgrad(fakeqm, oldgrd_qm);
    ommp_full_geomgrad(sys, oldgrd_mm);
    for(int i=0; i < 3 * qm_atoms; i++) qmg[i] += oldgrd_qm[i];
    for(int i=0; i < 3 * mm_atoms; i++) mmg[i] += oldgrd_mm[i];
    
    if(qmh != NULL) ommp_qm_helper_link_atom_geomgrad(qmh, sys, oldgrd_qm, oldgrd_mm, qmg);
    for(int i=0; i < 3 * qm_atoms; i++) qmg[i] += oldgrd_qm[i];
    for(int i=0; i < 3 * mm_atoms; i++) mmg[i] += oldgrd_mm[i];
    
    free(oldgrd_mm);
    free(oldgrd_qm);
}

int main(int argc, char **argv){
    if(argc != 3){
        printf("Given a JSON SmartInput file, it computes the geoemtrical derivatives \n");
        printf("with respect to the MM atoms coordinates using analytical derivatives.\n");
        printf("If a QM system is defined in the JSON file and atom types are provided, \n");
        printf("it is converted in a non-pol MM system interacting with the main one only \n");
        printf("generating an electric field that is used for induced dipole calculation \n");
        printf("(but not in electrostatic interaction!), with VdW interaction and, if \n");
        printf("defined, by link-atoms.\n\n");
        printf("Syntax expected\n");
        printf("    $ test_SI_geomgrad.exe <JSON FILE> <OUTPUT FILE>\n");
        return 1;
    }
    
    OMMP_SYSTEM_PRT my_system, fake_qm = NULL;
    OMMP_QM_HELPER_PRT my_qmh = NULL;
    ommp_smartinput(argv[1], &my_system, &my_qmh);
    ommp_set_outputfile(argv[2]);
    
    bool use_qm = false, use_fake_qm = false;
    
    if(my_qmh != NULL){
        // A QM part is present!
        use_qm = true;

        if(ommp_use_linkatoms(my_system)){
            // Since LA are used, I have to create a 
            // fake system for QM in order to have the
            // complete energy; parameters should be set 
            // anyway.
            use_fake_qm = true;
        }
    }

    if(use_fake_qm){
        // Cherrypick the parameter file use for QM from
        // the smart input
        char *prm_file, addr[] = "qm/prm_file/path";
        ommp_smartinput_cpstr(argv[1], addr, &prm_file);
        //Create the fake qm system
        fake_qm = ommp_system_from_qm_helper(my_qmh, prm_file);
        // In order to make a safe interaction, remove all the polarizabilities
        // on the fake qm system
        int natm = ommp_get_pol_atoms(fake_qm);
        int32_t *nopol = malloc(sizeof(int32_t) * natm);
        for(int j=0; j < natm; j++)
            nopol[j] = j;
        ommp_turn_pol_off(fake_qm, natm, nopol);
        free(nopol);
    }

    ana_grd_print(my_system, ommp_fixedelec_geomgrad, "EM");
    ana_grd_print(my_system, ommp_polelec_geomgrad, "EP");
    ana_grd_print(my_system, ommp_vdw_geomgrad, "EV");
    ana_grd_print(my_system, ommp_bond_geomgrad, "EB");
    ana_grd_print(my_system, ommp_angle_geomgrad, "EA");
    ana_grd_print(my_system, ommp_strbnd_geomgrad, "EBA");
    ana_grd_print(my_system, ommp_urey_geomgrad, "EUB");
    ana_grd_print(my_system, ommp_opb_geomgrad, "EOPB");
    ana_grd_print(my_system, ommp_pitors_geomgrad, "EPT");
    ana_grd_print(my_system, ommp_torsion_geomgrad, "ET");
    ana_grd_print(my_system, ommp_tortor_geomgrad, "ETT");
    ana_grd_print(my_system, ommp_imptorsion_geomgrad, "EIT");
    ana_grd_print(my_system, ommp_strtor_geomgrad, "EBT");
    ana_grd_print(my_system, ommp_angtor_geomgrad, "EAT");
    if(use_qm){
        ana_grd_print_qmmm(my_system, my_qmh, fake_qm,
                           ommptest_qm_helper_vdw_geomgrad, "EVQMMM");
        if(use_fake_qm){
            ana_grd_print_qmmm(my_system, my_qmh, fake_qm,
                               ommptest_fakeqm_internal_geomgrad, "EQM");
            ana_grd_print_qmmm(my_system, my_qmh, fake_qm,
                               ommptest_fakeqm_linkatom_geomgrad, "ELA");
        }
        ana_grd_print_qmmm(my_system, my_qmh, fake_qm,
                            ommptest_totalqmmm_geomgrad, "ETOT");
    }
    else{
        ana_grd_print(my_system, ommp_full_geomgrad, "ETOT");
    }
    
    if(my_qmh != NULL) ommp_terminate_qm_helper(my_qmh);
    ommp_terminate(my_system);
    
    return 0;
}

File: src/test_programs/F03/test_SI_geomgrad.f90
module test_geomgrad
    use ommp_interface

    abstract interface
    subroutine grad_term(s, grad)
        use mod_mmpol, only: ommp_system
        use mod_memory, only: rp
        type(ommp_system), intent(inout), target :: s
        real(rp), intent(out) :: grad(3,s%top%mm_atoms)
    end subroutine
    end interface
    
    abstract interface
    subroutine grad_term_qmmm(sys, qmh, fakeqm, gmm, gqm)
        use mod_mmpol, only: ommp_system
        use mod_qm_helper, only: ommp_qm_helper
        use mod_memory, only: rp
        type(ommp_system), intent(inout), target :: sys, fakeqm
        type(ommp_qm_helper), intent(inout), target :: qmh
        real(rp), intent(out) :: gmm(3,sys%top%mm_atoms), gqm(3,qmh%qm_top%mm_atoms)
    end subroutine
    end interface
    
    contains

    subroutine ommptest_fakeqm_internal_geomgrad(sys, qmh, fakeqm, gmm, gqm)
        type(ommp_system), intent(inout), target :: sys, fakeqm
        type(ommp_qm_helper), intent(inout), target :: qmh
        real(ommp_real), intent(out) :: gmm(3,sys%top%mm_atoms), gqm(3,qmh%qm_top%mm_atoms)

        gmm = 0.0
        call ommp_full_geomgrad(fakeqm, gqm)
    end subroutine

    subroutine ommptest_fakeqm_linkatom_geomgrad(sys, qmh, fakeqm, gmm, gqm)
        type(ommp_system), intent(inout), target :: sys, fakeqm
        type(ommp_qm_helper), intent(inout), target :: qmh
        real(ommp_real), intent(out) :: gmm(3,sys%top%mm_atoms), gqm(3,qmh%qm_top%mm_atoms)
        
        real(ommp_real), allocatable, dimension(:,:) :: tmp
        allocate(tmp(3, qmh%qm_top%mm_atoms))
        tmp = 0.0
        call ommp_qm_helper_vdw_geomgrad(qmh, sys, gqm, gmm)
        call ommp_full_geomgrad(fakeqm, tmp)
        tmp = tmp + gqm
        call ommp_qm_helper_link_atom_geomgrad(qmh, sys, gqm, gmm, tmp);
        deallocate(tmp)
    end subroutine

    subroutine ommptest_qm_helper_vdw_geomgrad(sys, qmh, fakeqm, gmm, gqm)
        type(ommp_system), intent(inout), target :: sys, fakeqm
        type(ommp_qm_helper), intent(inout), target :: qmh
        real(ommp_real), intent(out) :: gmm(3,sys%top%mm_atoms), gqm(3,qmh%qm_top%mm_atoms)

        call ommp_qm_helper_vdw_geomgrad(qmh, sys, gqm, gmm)
    end subroutine

    subroutine ommptest_totalqmmm_geomgrad(sys, qmh, fakeqm, gmm, gqm)
        type(ommp_system), intent(inout), target :: sys, fakeqm
        type(ommp_qm_helper), intent(inout), target :: qmh
        real(ommp_real), intent(out) :: gmm(3,sys%top%mm_atoms), gqm(3,qmh%qm_top%mm_atoms)

        real(ommp_real), allocatable, dimension(:,:) :: tmpmm, tmpqm

        gmm = 0.0
        gqm = 0.0
        allocate(tmpmm(3, sys%top%mm_atoms))
        tmpmm = 0.0
        allocate(tmpqm(3, qmh%qm_top%mm_atoms))
        tmpqm = 0.0

        call ommp_qm_helper_vdw_geomgrad(qmh, sys, tmpqm, tmpmm)
        gmm = gmm + tmpmm
        gqm = gqm + tmpqm
        if(sys%use_linkatoms) then
            call ommp_full_geomgrad(fakeqm, tmpqm)
            gqm = gqm + tmpqm
        end if
        call ommp_full_geomgrad(sys, tmpmm)
        gmm = gmm + tmpmm
        call ommp_qm_helper_link_atom_geomgrad(qmh, sys, tmpqm, tmpmm, gqm)
        gmm = gmm + tmpmm
        gqm = gqm + tmpqm
        deallocate(tmpmm, tmpqm)
    end subroutine

    subroutine print_qmmm_grad(n, mmat, qmat, mmg, qmg)
        character(len=*) :: n
        integer(ommp_integer) :: mmat, qmat
        real(ommp_real) :: mmg(3,mmat), qmg(3,qmat)

        character(len=OMMP_STR_CHAR_MAX) :: msg
        integer(ommp_integer) :: i
        write(msg, "('Grad ', A)") n
        call ommp_message(msg, OMMP_VERBOSE_NONE, "TEST-GRD")

        do i=1, mmat
            write(msg, "('MM:', I0, E20.12, ' ', E20.12, ' ', E20.12)") &
                i, mmg(:,i)
            call ommp_message(msg, OMMP_VERBOSE_NONE, "TEST-GRD")
        end do

        if(qmat > 0) then
            do i=1, qmat
                write(msg, "('QM:', I08, E20.12, ' ', E20.12, ' ', E20.12)") &
                    i, qmg(:,i)
                call ommp_message(msg, OMMP_VERBOSE_NONE, "TEST-GRD")
            end do
        end if

        call ommp_message("", OMMP_VERBOSE_NONE, "TEST-GRD")
    end subroutine

    subroutine ana_grd_print(sys, grad_f, n)
        type(ommp_system) :: sys
        procedure(grad_term), pointer :: grad_f
        character(len=*) :: n
        
        real(ommp_real), allocatable, dimension(:,:) :: gmm
        real(ommp_real), dimension(3,0) :: gqm

        allocate(gmm(3, sys%top%mm_atoms))
        gmm = 0.0
        call grad_f(sys, gmm)
        gmm = gmm * OMMP_AU2KCALMOL*OMMP_ANG2AU
        call print_qmmm_grad(n, sys%top%mm_atoms, 0, gmm, gqm)
        deallocate(gmm)

    end subroutine

    subroutine ana_grd_print_qmmm(sys, qmh, fakeqm, grad_f, n)
        type(ommp_system) :: sys, fakeqm
        type(ommp_qm_helper) :: qmh
        procedure(grad_term_qmmm), pointer :: grad_f
        character(len=*) :: n
        
        real(ommp_real), allocatable, dimension(:,:) :: gmm
        real(ommp_real), allocatable, dimension(:,:) :: gqm

        allocate(gmm(3, sys%top%mm_atoms))
        gmm = 0.0
        allocate(gqm(3, qmh%qm_top%mm_atoms))
        gqm = 0.0
        call grad_f(sys, qmh, fakeqm, gmm, gqm)
        gmm = gmm * OMMP_AU2KCALMOL*OMMP_ANG2AU
        gqm = gqm * OMMP_AU2KCALMOL*OMMP_ANG2AU
        call print_qmmm_grad(n, sys%top%mm_atoms, qmh%qm_top%mm_atoms, gmm, gqm)
        deallocate(gmm)
        deallocate(gqm)

    end subroutine
end module

program test_SI_geomgrad
    use iso_c_binding, only: c_char
    use ommp_interface
    use test_geomgrad

    implicit none
    character(kind=c_char, len=120), dimension(3) :: args
    character(len=OMMP_STR_CHAR_MAX) :: prm_file
    integer :: narg
    type(ommp_system), pointer :: my_system, fake_qm
    type(ommp_qm_helper), pointer :: my_qmh
    logical :: use_qm = .false., use_fake_qm = .false.

    procedure(grad_term), pointer :: gt
    procedure(grad_term_qmmm), pointer :: gt_qmmm
  
    narg = command_argument_count()
    if (narg /= 2 .and. narg /= 1) then
        write(6, *) "Syntax expected "
        write(6, *) "   $ test_SI_geomgrad.exe <JSON FILE> <OUTPUT FILE>"
        call exit(1)
    else 
        call get_command_argument(1, args(1))
        call get_command_argument(2, args(2))
       
        call ommp_smartinput(trim(args(1)), my_system, my_qmh)
        call ommp_set_outputfile(trim(args(2)))

        if(associated(my_qmh)) then
            use_qm = .true.
            if(my_system%use_linkatoms) then
                use_fake_qm = .true.
                call ommp_smartinput_cpstr(trim(args(1)), "qm/prm_file/path", prm_file);
                call ommp_system_from_qm_helper(my_qmh, trim(prm_file), fake_qm);
                call ommp_turn_pol_off(fake_qm, fake_qm%top%mm_atoms, fake_qm%eel%polar_mm)
            else
                allocate(fake_qm)
            end if
        end if
        
        gt => ommp_fixedelec_geomgrad
        call ana_grd_print(my_system, gt, "EM")
        gt => ommp_polelec_geomgrad
        call ana_grd_print(my_system, gt, "EP");
        gt => ommp_vdw_geomgrad
        call ana_grd_print(my_system, gt, "EV");
        gt => ommp_bond_geomgrad
        call ana_grd_print(my_system, gt, "EB");
        gt => ommp_angle_geomgrad
        call ana_grd_print(my_system, gt, "EA");
        gt => ommp_strbnd_geomgrad
        call ana_grd_print(my_system, gt, "EBA");
        gt => ommp_urey_geomgrad
        call ana_grd_print(my_system, gt, "EUB");
        gt => ommp_opb_geomgrad
        call ana_grd_print(my_system, gt, "EOPB");
        gt => ommp_pitors_geomgrad
        call ana_grd_print(my_system, gt, "EPT");
        gt => ommp_torsion_geomgrad
        call ana_grd_print(my_system, gt, "ET");
        gt => ommp_tortor_geomgrad
        call ana_grd_print(my_system, gt, "ETT");
        gt => ommp_imptorsion_geomgrad
        call ana_grd_print(my_system, gt, "EIT");
        gt => ommp_strtor_geomgrad
        call ana_grd_print(my_system, gt, "EBT");
        gt => ommp_angtor_geomgrad
        call ana_grd_print(my_system, gt, "EAT");
        if(use_qm) then
            gt_qmmm => ommptest_qm_helper_vdw_geomgrad
            call ana_grd_print_qmmm(my_system, my_qmh, fake_qm, gt_qmmm, "EVQMMM")
            if(use_fake_qm) then
                gt_qmmm => ommptest_fakeqm_internal_geomgrad
                call ana_grd_print_qmmm(my_system, my_qmh, fake_qm, gt_qmmm, "EQ")
                gt_qmmm => ommptest_fakeqm_linkatom_geomgrad
                call ana_grd_print_qmmm(my_system, my_qmh, fake_qm, gt_qmmm, "ELA")
            end if
            gt_qmmm => ommptest_totalqmmm_geomgrad
            call ana_grd_print_qmmm(my_system, my_qmh, fake_qm, gt_qmmm, "ETOT")
        else
            gt => ommp_full_geomgrad
            call ana_grd_print(my_system, gt, "ETOT")
        end if

        if(associated(my_qmh)) call ommp_terminate_qm_helper(my_qmh)
        if(associated(my_system)) call ommp_terminate(my_system)
    end if
end program
